### PR TITLE
Fix MySQL isolation level for read-modify-write ops

### PR DESCRIPTION
**Pull request by azdagron**
Certain operations in SPIRE are read-modify-write, i.e., they read rows, process columns on those rows, and then update the rows as part of a single transaction. This is for sensitive operations like bundle appending/pruning.

We've relied on the REPEATABLE READ isolation level to provide us the consistency guarantees we need over these kind of operations operation, namely that which prevents concurrent transactions from updating the same target row(s). This is to prevent one transaction from undoing the action just performed by another. For example, if two servers attempt to append key material to the bundle concurrently, we don't want the append op from one to accidentally "undo" the append of the other, as they are both updating the same bundle row.

We chose REPEATABLE READ based on an assumption that PostgreSQL behavior was universal.

Unfortunately MySQL REPEATABLE READ is weaker than that of PostgreSQL. While, PostgreSQL fails concurrent transactions that try to update the same target row, MySQL does not. For MySQL, we need SERIALIZABLE, which is the same as REPEATABLE READ except that it converts `SELECT` to `SELECT ... LOCK FOR SHARE MODE` which "sets a shared lock that permits other transactions to read the examined rows but not to update or delete them", which gives us the result we need.

Fixes: #2149

### CODE REVIEW

File: pkg/server/plugin/datastore/sql/sql.go [hunk 1]
```diff
@@ -561,14 +561,38 @@ func (*Plugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.G
...
+// operations that will read one or more rows, change one or more columns in
+// those rows, and then set them back. This requires a stronger level of
+// consistency that prevents two transactions from doing read-modify-write
+// concurrently.
+func (ds *Plugin) withReadModifyWriteTx(ctx context.Context, op func(tx *gorm.DB) error) error {
+	isolationLevel := sql.LevelRepeatableRead
+	if ds.db.databaseType == MySQL {
+		// MySQL REPEATABLE READ is weaker than that of PostgreSQL. Namely,
+		// PostgreSQL, beyond providing the minimum consistency guarantees
+		// mandataded for REPEATABLE READ in the standard, automatically fails
```
Comment (rturner3) [inline pkg/server/plugin/datastore/sql/sql.go:574]: nit: spelling of "mandated"

File: pkg/server/plugin/datastore/sql/sql.go [hunk 2]
```diff
@@ -561,14 +561,38 @@ func (*Plugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.G
...
-
+// withReadModifyWriteTx wraps the operation in a transaction appropriate for
+// operations that will read one or more rows, change one or more columns in
+// those rows, and then set them back. This requires a stronger level of
+// consistency that prevents two transactions from doing read-modify-write
+// concurrently.
+func (ds *Plugin) withReadModifyWriteTx(ctx context.Context, op func(tx *gorm.DB) error) error {
+	isolationLevel := sql.LevelRepeatableRead
+	if ds.db.databaseType == MySQL {
+		// MySQL REPEATABLE READ is weaker than that of PostgreSQL. Namely,
```
Comment (rturner3) [inline pkg/server/plugin/datastore/sql/sql.go:572]: Thanks for adding all these comments! Really helps to give the context

**Review by rturner3 — COMMENTED** [review]
I think we might need to do the same thing for updating of attested nodes

**Comment by azdagron** [comment]
Good catch! I've updated the PR.

**Comment by azdagron** [comment]
Github is struggling. Commit is not reflected here yet.

**Review by rturner3 — APPROVED** [review]
🚀

**Pull request merged**